File: ol/Ordered list/Ordered list.cs
using System;
using System.Collections.Generic;
using System.Runtime.ExceptionServices;
using System.Security.Policy;
using System.Xml.Linq;

namespace AlgorithmsDataStructures
{

    public class Node<T>
    {
        public T value;
        public Node<T> next, prev;

        public Node(T _value)
        {
            value = _value;
            next = null;
            prev = null;
        }
    }

    public class OrderedList<T>
    {
        public Node<T> head, tail;
        private bool _ascending;

        public OrderedList(bool asc)
        {
            head = null;
            tail = null;
            _ascending = asc;
        }

        public int Compare(T v1, T v2)
        {
            if (typeof(T) == typeof(String))
            {
                string firstNumber = Convert.ToString(v1).Trim();
                string secondNumber = Convert.ToString(v2).Trim();
                return firstNumber.CompareTo(secondNumber);
            }
            else
            {
                int firstNumber = Convert.ToInt32(v1);
                int secondNumber = Convert.ToInt32(v2);
                return firstNumber.CompareTo(secondNumber);
            }
        }

        public void Add(T value)
        {
            Node<T> node = head;
            Node<T> insertNode = new Node<T>(value);

            if (_ascending) AscendingInsert(node, insertNode);
            else DescendingInsert(node, insertNode);
        }

        public Node<T> Find(T val)
        {
            if (head == null) return default;
            if (_ascending && Compare(head.value, val) > 0) return default;
            if (_ascending && Compare(tail.value, val) < 0) return default;
            if (!_ascending && Compare(head.value, val) < 0) return default;
            if (!_ascending && Compare(tail.value, val) > 0) return default;

            Node<T> node = head;
            while (node != null)
            {
                if (Compare(node.value, val) == 0) return node;
                node = node.next;
            }
            return default;
        }

        public void Delete(T val)
        {
            Node<T> node = Find(val);
            bool isFirst = node == head;
            bool isLast = node == tail;

            if (isFirst && isLast)
            {
                head = tail = null;
            }
            else if (isFirst)
            {
                head = head.next;
                head.prev = null;
            }            
            else if (isLast)
            {
                tail = tail.prev;
                tail.next = null;
            }
            else if (node != default)
            {
                node.prev.next = node.next;
                node.next.prev = node.prev;
                node.next = node.next.next;
            }
        }

        public void Clear(bool asc)
        {
            _ascending = asc;

            Node<T> node = head;

            while (node != null)
            {
                head = head.next;
                node.next = null;
                node = head;

                if (node == null) tail = null;
                else head.prev = null;
            }
        }

        public int Count()
        {
            int length = 0;
            Node<T> node = head;
            
            while (node != null)
            {
                ++length;
                node = node.next;
            }
            return length;
        }

        public List<Node<T>> GetAll()
        {
            List<Node<T>> r = new List<Node<T>>();
            Node<T> node = head;
            while (node != null)
            {
                r.Add(node);
                node = node.next;
            }
            return r;
        }

        private void AscendingInsert(Node<T> compareNode, Node<T> insertNode)
        {
            int comparison = 0;
            
            while (compareNode != null)
            {
                comparison = Compare(insertNode.value, compareNode.value);

                if (comparison > 0 && compareNode.next is null)
                {
                    compareNode.next = insertNode;
                    insertNode.prev = compareNode;
                    insertNode.next = null;
                    tail = insertNode;
                    break;
                }
                else if (comparison < 1 && compareNode.prev is null)
                {
                    compareNode.prev = insertNode;
                    insertNode.next = compareNode;
                    insertNode.prev = null;
                    head = insertNode;
                    break;
                } 
                else if (comparison < 1)
                {
                    insertNode.prev = compareNode.prev;
                    compareNode.prev.next = insertNode;
                    compareNode.prev = insertNode;
                    insertNode.next = compareNode;
                    break;
                }
                compareNode = compareNode.next;
            }

            if (insertNode.prev is null) head = insertNode;
            if (insertNode.next is null) tail = insertNode;
        }

        private void DescendingInsert(Node<T> compareNode, Node<T> insertNode)
        {
            int comparison = 0;
            
            while (compareNode != null)
            {
                comparison = Compare(insertNode.value, compareNode.value);

                if (comparison < 0 && compareNode.next is null)
                {
                    compareNode.next = insertNode;
                    insertNode.prev = compareNode;
                    insertNode.next = null;
                    tail = insertNode;
                    break;
                }
                else if (comparison > -1 && compareNode.prev is null)
                {
                    compareNode.prev = insertNode;
                    insertNode.next = compareNode;
                    insertNode.prev = null;
                    head = insertNode;
                    break;
                }
                else if (comparison > -1)
                {
                    insertNode.prev = compareNode.prev;
                    compareNode.prev.next = insertNode;
                    compareNode.prev = insertNode;
                    insertNode.next = compareNode;
                    break;
                }
                compareNode = compareNode.next;
            }

            if (insertNode.prev is null) head = insertNode;
            if (insertNode.next is null) tail = insertNode;
        }
    }
}

File: 07.ordered list/Ordered list/Python/ordered_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.prev = None
        self.next = None


class OrderedList:
    def __init__(self, asc):
        self.head = None
        self.tail = None
        self.__ascending = asc

    def compare(self, v1, v2):
        # -1 если v1 < v2
        # 0 если v1 == v2
        # +1 если v1 > v2
        if v1 < v2:
            return -1
        elif v1 == v2:
            return 0
        else:
            return +1

    def add(self, value):
        # автоматическая вставка value
        # в нужную позицию
        node = self.head
        old = None

        while node is not None:
            if self.__ascending and self.compare(node.value, value) == +1:
                break
            if not self.__ascending and self.compare(node.value, value) == -1:
                break
            old = node
            node = node.next

        newNode = Node(value)
        if old is None:
            if self.tail is None:
                self.head = newNode
                self.tail = newNode
            else:
                newNode.next = self.head
                self.head.prev = newNode
                self.head = newNode
        else:
            newNode.next = node
            newNode.prev = old
            old.next = newNode
            if node is None:
                self.tail = newNode
            else:
                node.prev = newNode

    def find(self, val):
        node = self.head
        stop = False
        while node is not None and not stop:
            if node.value == val:
                return node
            else:
                if self.__ascending and node.value > val or \
                        not self.__ascending and node.value < val:
                    stop = True
                else:
                    node = node.next
        return None

    def delete(self, val):
        if self.head is None:
            return

        node = self.head

        while node is not None and node.value == val:
            self.head = self.head.next
            # node = self.head
            if self.head is None:
                self.tail = None
            else:
                self.head.prev = None
            return

        while node is not None:
            while node is not None and node.value != val:
                node = node.next
            if node is None:
                return
            old = node
            old.prev.next = node.next
            if node == self.tail:
                self.tail = self.tail.prev
                self.tail.next = None
            else:
                old.next.prev = node.prev
            return

    def clean(self, asc):
        if self.head is None:
            return
        node = self.head
        while node is not None:
            self.head = self.head.next
            node.value = None
            node.next = None
            node = self.head
            if node is None:
                self.tail = None
            else:
                self.head.prev = None

        self.__ascending = asc

    def len(self):
        if self.head is None:
            return 0
        node = self.head
        length = 0
        while node is not None:
            node = node.next
            length = length + 1
        return length

    def get_all(self):
        r = []
        node = self.head
        while node != None:
            r.append(node)
            node = node.next
        return r

    def print_all_nodes(self):
        node = self.head
        while node is not None:
            print(node.value)
            node = node.next


class OrderedStringList(OrderedList):
    def __init__(self, asc):
        super(OrderedStringList, self).__init__(asc)

    def compare(self, v1, v2):
        if ''.join(v1.split()) < ''.join(v2.split()):
            return -1
        elif ''.join(v1.split()) == ''.join(v2.split()):
            return 0
        else:
            return +1
